In [1]:
# requests from quotes.py not working so getting quotes here after downloading .html files from website manually
from bs4 import BeautifulSoup
from datetime import datetime, date, timedelta
import json
import unicodedata
import pprint
import re

### 2025

In [72]:
# 2025
with open('html files/365 Best Inspirational Quotes for 2025.html') as f:
    soup = BeautifulSoup(f, 'html.parser')

# Find all date tags
dates = soup.find_all('b')

# Initialize a list to store the data
quotes = []

for date in dates:
    date_text = date.text.strip()
    # %A = day of week full name e.g. Sunday
    # #B = month full name e.g. January
    # %d = day of month with zero padding
    # %-d = day as decimal number i.e. without zero padding
    # %Y = full year e.g. 2025
    date_format = "%A %B %d, %Y"
    try:
        parsed_date = datetime.strptime(date_text, date_format).strftime("%Y-%m-%d")

        quote = date.next_sibling.next_sibling
        quote_text = quote.strip()[1:-1]

        author = quote.next_sibling.next_sibling[1:].strip()

        quotes.append({"date": parsed_date, 
                        "quote": quote_text,
                        "author": author})
    except ValueError:
        # Skip if the date format doesn't match
        continue

# Write the data to a JSON file
with open('quotes_2025.json', 'w') as f:
    json.dump(quotes, f, indent=4)

### 2024

In [ ]:
# 2024
with open('html files/366 Top Inspirational Quotes for 2024.html') as f:
    soup = BeautifulSoup(f)

    # Find all date tags
    dates = soup.find_all('h2')

    # Initialize a list to store the data
    quotes = []

    for date in dates:
        date_text = date.text.strip()
        # %A = day of week full name e.g. Sunday
        # #B = month full name e.g. January
        # %d = day of month with zero padding
        # %-d = day as decimal number i.e. without zero padding
        # %Y = full year e.g. 2025
        date_format = "%A, %B %d, %Y"
        try:
            parsed_date = datetime.strptime(date_text, date_format).strftime("%Y-%m-%d")

            contents = date.next_sibling.next_sibling.contents

            quote = contents[0].text.strip()[1:-1]
            author = contents[3].text.strip()

            quotes.append({"date": parsed_date, 
                            "quote": quote,
                            "author": author})
        except ValueError:
            # Skip if the date format doesn't match
            continue

with open('quotes_2024.json', 'w') as f:
    json.dump(quotes, f, indent=4)

### 2023

In [ ]:
with open('html files/365 Best Inspirational Quotes for 2023.html') as f:
    soup = BeautifulSoup(f)

    # Find all date tags
    dates = soup.find_all('h2')

    # Initialize a list to store the data
    quotes = []

    for date in dates:
        date_text = date.text.strip()
        # %A = day of week full name e.g. Sunday
        # #B = month full name e.g. January
        # %d = day of month with zero padding
        # %-d = day as decimal number i.e. without zero padding
        # %Y = full year e.g. 2025
        date_format = "%B %d, %Y (%A)"
        try:
            parsed_date = datetime.strptime(date_text, date_format).strftime("%Y-%m-%d")

            quote = date.find_next("p").find_next("p")
            author = quote.next.next.next.text.strip()[1:].strip()
            quote = quote.next.text.strip()[1:-1]

            quotes.append({"date": parsed_date, 
                            "quote": quote,
                            "author": author})
        except ValueError:
            # Skip if the date format doesn't match
            continue

In [ ]:
# figure out which quotes are missing and manually add them
json_dates = set()
for quote in quotes:
    json_dates.add(quote['date'])    

# https://stackoverflow.com/a/66595046/11065894
start_date = date(2023, 1, 1)
end_date = date(2023, 12, 31)

dates_2023 = {((start_date + timedelta(days=x))).strftime("%Y-%m-%d") for x in range((end_date + timedelta(days=1) - start_date).days)}

dates_2023 - json_dates

In [94]:
quotes.append({"date": '2023-01-11',
               "quote": "Certain things in life are more important than the usual crap that everyone strives for.",
               "author": "James Taylor"
})

quotes.append({"date": '2023-03-16',
    "quote": "The man who moves a mountain begins by carrying away small stones.",
    "author": "Confucius"
})
quotes.append({"date": '2023-03-17',
               "quote": "When 99% of people doubt your idea, you’re either gravely wrong or about to make history.",
"author": "Scott Belsky"
})

quotes.append({"date": '2023-03-18',
               "quote": "Success is a lot like a bright, white tuxedo. You feel terrific when you get it, but then you’re desperately afraid of getting it dirty, of spoiling it in any way.",
"author": "Conan O'Brien"
})

quotes.append({"date": '2023-03-19',
               "quote": "I get tired of myself pretty fast. So I can’t imagine other people don’t get tired of me.",
"author": "Paul Giamatti"
})

quotes.append({"date": '2023-03-20',
               "quote": "If there is no struggle, there is no progress.",
"author": "Frederick Douglass"
})

quotes.append({"date": '2023-03-21',
               "quote": "I think people waste so much time strategizing about what they should do, rather than just going and doing something, making mistakes, and then pivoting. The goal of a startup should be: Launch as many things as possible, fail as quickly as possible, and then figure out how to move forward from there.",
"author": "Jenny Fleiss"
})

with open('quotes_2023.json', 'w') as f:
    json.dump(quotes, f, indent=4)

### 2022

In [15]:
with open('html files/365 Inspirational Quotes for 2022.html') as f:
    soup = BeautifulSoup(f)

    # Find all date tags
    dates = soup.find_all('h2')

    # Initialize a list to store the data
    quotes = []

    for date in dates:
        date_text = date.text.strip()
        # %A = day of week full name e.g. Sunday
        # #B = month full name e.g. January
        # %d = day of month with zero padding
        # %-d = day as decimal number i.e. without zero padding
        # %Y = full year e.g. 2025
        date_format = "%A, %B %d, %Y"
        try:
            parsed_date = datetime.strptime(date_text, date_format).strftime("%Y-%m-%d")

            quote = date.find_next("p")
            author = quote.find_next("em").text.strip()
            quote = quote.next.text.strip()[1:-1]

            quotes.append({"date": parsed_date, 
                            "quote": quote,
                            "author": author})
        except ValueError:
            # Skip if the date format doesn't match
            continue

with open('quotes_2022.json', 'w') as f:
    json.dump(quotes, f, indent=4)

### 2019

In [2]:
with open('html files/Here Are the Best Inspirational Quotes for 2019 _ Inc.com.html') as f:
    soup = BeautifulSoup(f)

    dates = soup.find_all('h2')

    quotes = []
    for idx, date in enumerate(dates):
        date_text = date.text.split("(")[0].strip()
        # %A = day of week full name e.g. Sunday
        # #B = month full name e.g. January
        # %d = day of month with zero padding
        # %-d = day as decimal number i.e. without zero padding
        # %Y = full year e.g. 2025
        date_format = "%A, %B %d, %Y"
        try:
            parsed_date = datetime.strptime(date_text, date_format).strftime("%Y-%m-%d")
        except ValueError:
            print(f"error: date format doesn't match:\n{date_text}")
            continue

        quote = date.find_next("p")
        author = quote.find_next('br').next.text[2:].strip()

        quote_text = unicodedata.normalize("NFKC", quote.next.text)
        if quote_text.startswith('"') and quote_text.endswith('"'):
            quote_text = quote_text[1: -1]
        else:
            print(idx, date, quote_text)

        quotes.append({"date": parsed_date, 
                            "quote": quote_text,
                            "author": author})

24 <h2>Friday, January 25, 2019</h2> "The last 10 percent it takes to launch something takes as much energy as the first 90 percent." --Rob Kalin
25 <h2>Saturday, January 26, 2019</h2> I urge you to please notice when you are happy, and exclaim or murmur or think at some point, 'If this isn't nice, I don't know what is.'"
96 <h2>Sunday, April 7, 2019</h2> "I have learned over the years that when one's mind is made up, this diminishes fear."        --Rosa Parks
110 <h2>Sunday, April 21, 2019 (Easter Sunday)</h2> Pray as though everything depended on God. Work as though everything depended on you."
117 <h2>Sunday, April 28, 2019</h2> The crisis of today is the joke of tomorrow."
159 <h2>Sunday, June 9, 2019</h2> "Live daringly, boldly, fearlessly. Taste the relish to be found in competition
186 <h2>Saturday, July 6, 2019</h2> "I hate that word: 
356 <h2>Wednesday, December 25, 2019</h2> Christmas
error: date format doesn't match:
Bonus quote


In [3]:
quotes[24] = {'date': '2019-01-25',
 'quote': 'The last 10 percent it takes to launch something takes as much energy as the first 90 percent.',
 'author': 'Rob Kalin'}

quotes[25] = {'date': '2019-01-26',
 'quote': 'I urge you to please notice when you are happy, and exclaim or murmur or think at some point, \'If this isn\'t nice, I don\'t know what is.',
 'author': 'Kurt Vonnegut'}

quotes[96] = {'date': '2019-04-07',
 'quote': 'I have learned over the years that when one\'s mind is made up, this diminishes fear.',
 'author': 'Rosa Park'}

quotes[110] = {'date': '2019-04-21',
 'quote': 'Pray as though everything depended on God. Work as though everything depended on you.',
 'author': 'Saint Augustine'}

quotes[117] = {'date': '2019-04-28',
 'quote': 'The crisis of today is the joke of tomorrow.',
 'author': 'H.G. Wells'}

quotes[159] = {'date': '2019-06-09',
 'quote': 'Live daringly, boldly, fearlessly. Taste the relish to be found in competition--in having put forth the best within you.',
 'author': 'Henry J. Kaiser'}

quotes[186] = {'date': '2019-07-06',
 'quote': 'I hate that word: lucky. It cheapens a lot of hard work.',
 'author': 'Peter Dinklage'}

quotes[356] = {'date': '2019-12-25',
 'quote': 'There are two ways of spreading light: to be the candle or the mirror that reflects it.',
 'author': 'Edith Wharton'}

In [4]:
with open('quotes_2019.json', 'w') as f:
    json.dump(quotes, f, indent=4)

In [11]:
len(quotes)

363

In [9]:
# figure out which quotes are missing and manually add them
json_dates = set()
for quote in quotes:
    json_dates.add(quote['date'])    

# https://stackoverflow.com/a/66595046/11065894
start_date = date(2019, 1, 1)
end_date = date(2019, 12, 31)

dates_2019 = {((start_date + timedelta(days=x))).strftime("%Y-%m-%d") for x in range((end_date + timedelta(days=1) - start_date).days)}

dates_2019 - json_dates

{'2019-02-19', '2019-08-09', '2019-11-19'}

'2019-02-19' is missing because Feb 18 date was duplicated, **manually updated** in json file

'2019-08-09' and '2019-11-19' are missing

### 2020

In [2]:
with open('html files/366 Top Inspirational Quotes and Motivational Quotes for Every Single Day in 2020.html') as f:
    soup = BeautifulSoup(f)

    dates = soup.find_all('h2')

In [5]:
for idx, date in enumerate(dates):
    # print(date.find_next("div"))
    quote = date.find_next("p")
    quoteText = quote.next.text
    # author = quote.next.next.text[1:]
    author = quote.find_next('br').next.text[2:].strip()
    print(idx, ":", quoteText, "|", author)

0 : “I’ve learned that people will forget what you said, people will forget what you did, but people will never forget how you made them feel.”  | Maya Angelou
1 : “The thing about smart people is that they seem like crazy people to dumb people.” | Stephen Hawking
2 : “If you think you are too small to make a difference, try sleeping with a mosquito.” | Dalai Lama
3 : “Have no fear of perfection. You’ll never reach it.” | Salvador Dali
4 : “You can change. And you can be an agent of change.” | Laura Dern
5 : “Hard work and education will take you farther than any government program can ever promise.” | Mia Love
6 : “Wanting to be a good actor is not good enough. You must want to be a great actor. You just have to have that.” | Gary Oldman
7 : “When I first moved to New York, someone who thought they knew more than I did said: ‘You have to always look like you know where you’re going when you get out of the subway.'” | Maya Rudolph
8 : “It’s like soul music, isn’t it all soul music? Oth

In [6]:
with open('html files/366 Top Inspirational Quotes and Motivational Quotes for Every Single Day in 2020.html') as f:
    soup = BeautifulSoup(f)

    dates = soup.find_all('h2')

    quotes = []
    for idx, date in enumerate(dates):
        date_text = date.text.split("(")[0].strip()
        # %A = day of week full name e.g. Sunday
        # #B = month full name e.g. January
        # %d = day of month with zero padding
        # %-d = day as decimal number i.e. without zero padding
        # %Y = full year e.g. 2025
        date_format = "%A, %B %d, %Y"
        try:
            parsed_date = datetime.strptime(date_text, date_format).strftime("%Y-%m-%d")
        except ValueError:
            print(f"error: date format doesn't match:\n{date_text}")
            continue

        quote = date.find_next("p")
        author = quote.find_next('br').next.text[2:].strip()

        quote_text = quote.text.strip()[1: -1]
        quotes.append({"date": parsed_date, 
                            "quote": quote_text,
                            "author": author})
        
with open('quotes_2020.json', 'w') as f:
    json.dump(quotes, f, indent=4)

In [7]:
len(quotes)

366

### 2018

some of the html pages require pressing a button to show the rest of the quotes, worked around that by pressing the button and then saving the html after that

In [ ]:
with open('html files/Here Are the Best Inspirational Quotes for 2018 _ Inc.com.html') as f:
    soup = BeautifulSoup(f)

    dates = soup.find_all('h2')

In [24]:
dates[0].find_next('p').next.text

'"There\'s a tremendous bias against taking risks. Everyone is trying to optimize their ass-covering." --Elon Musk'

In [35]:
bad_strings = []
for idx, date in enumerate(dates):
    unpacked = date.find_next('p').next.text.split('--')
    if len(unpacked) > 2:
        print(idx, unpacked)
        bad_strings.append(date)


3 ['"Gifts are easy', 'they\'re given, after all. Choices can be hard. You can seduce yourself with your gifts if you\'re not careful, and if you do, it\'ll probably be to the detriment of your choices." ', 'Jeff Bezos']
9 ['"Don\'t be distracted by criticism. Remember', 'the only taste of success some people get is to take a bite out of you." ', 'Zig Ziglar']
24 ['"Today I choose life. To feel the freedom that comes from being able to continue to make mistakes and choices', 'today I choose to feel life, not to deny my humanity but embrace it." ', 'Kevyn Aucoin']
138 ['"Far better is it to dare mighty things, to win glorious triumphs ', ' even though checkered by failure ', ' than to rank with those poor spirits who neither enjoy much nor suffer much, because they live in a gray twilight that knows not victory nor defeat." ', 'Theodore Roosevelt']
163 ['"Keep your face always toward the sunshine', 'and shadows will fall behind you." ', 'Walt Whitman']
186 ['"The firsts go away', 'first

In [37]:
for date in bad_strings:
    print(date.find_next('p').next.text.rpartition('--'))

('"Gifts are easy--they\'re given, after all. Choices can be hard. You can seduce yourself with your gifts if you\'re not careful, and if you do, it\'ll probably be to the detriment of your choices." ', '--', 'Jeff Bezos')
('"Don\'t be distracted by criticism. Remember--the only taste of success some people get is to take a bite out of you." ', '--', 'Zig Ziglar')
('"Today I choose life. To feel the freedom that comes from being able to continue to make mistakes and choices--today I choose to feel life, not to deny my humanity but embrace it." ', '--', 'Kevyn Aucoin')
('"Far better is it to dare mighty things, to win glorious triumphs -- even though checkered by failure -- than to rank with those poor spirits who neither enjoy much nor suffer much, because they live in a gray twilight that knows not victory nor defeat." ', '--', 'Theodore Roosevelt')
('"Keep your face always toward the sunshine--and shadows will fall behind you." ', '--', 'Walt Whitman')
('"The firsts go away--first lo

In [41]:
with open('html files/Here Are the Best Inspirational Quotes for 2018 _ Inc.com.html') as f:
    soup = BeautifulSoup(f)

    dates = soup.find_all('h2')

    quotes = []
    for idx, date in enumerate(dates):
        date_text = date.text.split("(")[0].strip()
        # %A = day of week full name e.g. Sunday
        # #B = month full name e.g. January
        # %d = day of month with zero padding
        # %-d = day as decimal number i.e. without zero padding
        # %Y = full year e.g. 2025
        date_format = "%A, %B %d, %Y"
        try:
            parsed_date = datetime.strptime(date_text, date_format).strftime("%Y-%m-%d")
        except ValueError:
            try:
                parsed_date = datetime.strptime(date_text, "%A %B %d, %Y").strftime("%Y-%m-%d")
            except ValueError:
                print(f"error: date format doesn't match:\n{date_text}")
                continue

        quote, _, author = date.find_next("p").text.rpartition('--')

        author_text = author.strip()

        quote_text = quote.strip()[1: -1]
        quotes.append({"date": parsed_date, 
                            "quote": quote_text,
                            "author": author})
        
with open('quotes_2018.json', 'w') as f:
    json.dump(quotes, f, indent=4)

In [42]:
len(quotes)

365

### 2017

In [ ]:
skip_list = [27, 146, 184, 246, 304, 308, 309, 310, 335, 364]

with open('html files/365 Great Quotes for 2017 (Inspiring Words for the New Year).html') as f:
    soup = BeautifulSoup(f)

    # delete strong/br elements, most of holiday names are in strong
    for element in soup(['strong', 'br', 'a']):
        element.decompose()
    
    # delete empty tags
    for tag in soup.find_all():
        if not tag.text.strip() and not tag.contents:
            tag.decompose()

    dates = soup.find_all('h2')
    
    quotes = []
    
    for idx, date in enumerate(dates):

        date_text = date.text.split("(")[0].strip()
        # %A = day of week full name e.g. Sunday
        # #B = month full name e.g. January
        # %d = day of month with zero padding
        # %-d = day as decimal number i.e. without zero padding
        # %Y = full year e.g. 2025
        date_format = "%A, %B %d, %Y"
        try:
            parsed_date = datetime.strptime(date_text, date_format).strftime("%Y-%m-%d")
        except ValueError:
            try:
                parsed_date = datetime.strptime(date_text, "%A %B %d, %Y").strftime("%Y-%m-%d")
            except ValueError:
                print(f"error: date format doesn't match:\n{date_text}")
                continue

        quote = date.find_next('p').next
        quote_text = quote.text.strip()
        quote_text = quote_text.lstrip("“")
        quote_text = quote_text.rstrip("”")

        author = date.find_next('em').next.text.strip()[1:]

        author_text = author.strip()

        quotes.append({"date": parsed_date, 
                            "quote": quote_text,
                            "author": author_text})



In [302]:
len(quotes)

356

In [306]:
for idx in skip_list:
    print(dates[idx])

<h2>Saturday, January 28, 2017</h2>
<h2>Saturday, May 27, 2017</h2>
<h2>Tuesday, July 4, 2017</h2>
<h2>Monday, September 4, 2017</h2>
<h2>Wednesday, November 1, 2017</h2>
<h2>Sunday, November 5, 2017</h2>
<h2>Tuesday, November 7, 2017</h2>
<h2>Saturday, December 2, 2017</h2>
<h2>Sunday, December 31, 2017</h2>


In [316]:
quotes.append({'date': '2017-01-28',
 'quote': 'Everyone has a plan ’til they get punched in the mouth.',
 'author': 'Mike Tyson'})

quotes.append({'date': '2017-05-17',
 'quote': 'Sometimes you don’t realize how special something is until you lose it. Thats how I feel about Ramadan every year.',
 'author': 'Omar Suleiman'})

quotes.append({'date': '2017-06-04',
 'quote': 'America was not built on fear. America was built on courage, on imagination and an unbeatable determination to do the job at hand.',
 'author': 'Harry S. Truman'})

quotes.append({'date': '2017-09-04',
 'quote': 'I believe in the dignity of labor, whether with head or hand; that the world owes no man a living but that it owes every man an opportunity to make a living.',
 'author': 'John D. Rockefeller'})

quotes.append({'date': '2017-11-01',
 'quote': 'Pray as though everything depended on God. Work as though everything depended on you.',
 'author': 'Saint Augustine'})

quotes.append({'date': '2017-11-05',
 'quote': 'Time is what we want most, but what we use worst.',
 'author': 'William Penn'})

quotes.append({'date': '2017-11-06',
 'quote': 'There is no elevator to success — you have to take the stairs.',
 'author': 'Anonymous'})

quotes.append({'date': '2017-11-07',
 'quote': 'The vote is the most powerful instrument ever devised by man for breaking down injustice and destroying the terrible walls which imprison men because they are different from other men.',
 'author': 'Lyndon Johnson'})

quotes.append({'date': '2017-12-02',
 'quote': 'Human history begins with man’s act of disobedience which is at the very same time the beginning of his freedom and development of his reason.',
 'author': 'Erich Fromm'})

quotes.append({'date': '2017-12-31',
 'quote': 'An optimist stays up until midnight to see the new year in. A pessimist stays up to make sure the old year leaves.',
 'author': 'Bill Vaughn'})

In [317]:
with open('quotes_2017.json', 'w') as f:
    json.dump(quotes, f, indent=4)

In [318]:
len(quotes)

365

In [223]:
for idx, date in enumerate(dates):
    author = date.find_next('em').next.text.strip()
    if not author.startswith('–'):
        print(idx, date, author)

309 <h2>Monday, November 6, 2017</h2> — you have to take the stairs.”


In [246]:
skip_list = []
for idx, date in enumerate(dates):
    quote = date.find_next('p').next.text
    if not quote.startswith("“"):
        print(f"{idx}: {quote}")
        skip_list.append(idx)
skip_list

5: True luck consists not in holding the best of the cards at the table; luckiest is he who knows just when to rise and go home.”
27: Chinese New Year
146: Ramadan starts “Sometimes you don’t realize how special something is until you lose it. Thats how I feel about Ramadan every year.”
156: There on the beaches of Normandy I began to reflect on the wonders of these ordinary people whose lives were laced with the markings of greatness.”
184: Independence Day
246: Labor Day
304: All Saints’ Day
308: Daylight Saving Time ends
310: Election Day (New Jersey and Virginia governors)
335: International Day for the Abolition of Slavery
364: New Year’s Eve


[5, 27, 146, 156, 184, 246, 304, 308, 310, 335, 364]

In [247]:
skip_list.remove(5)
skip_list.remove(156)
skip_list

[27, 146, 184, 246, 304, 308, 310, 335, 364]

In [ ]:
quotes = []
keep_dates = []
pattern = r'“(.*?)”'
for idx, date in enumerate(dates):
    s = date.find_next('p').text
    match = re.search(pattern, s)
    if match:
        keep_dates.append(date)

        quotes.append(match.group(1))
    else:
        print(idx, s)

In [ ]:
for idx, date in enumerate(dates):    
    # pprint.pp(date.find_all_next()[:5])
    quote = date.find_next('p').next.text.strip()
    author = date.find_next('em')
    # print(f"{idx}: {quote=}, {author=}")
    if not quote.startswith('“') or not quote.endswith('”'):
        # quote = date.find_next('br').next.text
        print(f"{idx}, {quote=}, {author_text=}")

In [ ]:
for idx, date in enumerate(dates):
    quote = date.find_next('p').next.text
    author_text = date.find_next('em').next.text

    if not quote.startswith('“') or not quote.endswith('”'):
        quote = date.find_next('br').next.text
        print(f"{idx}, {quote=}, {author_text=}")

### 2016 & 2021 (partial)
I can't find all the quotes for each of these years as they are not in one page and apparently split into monthly installments, but can't find all pages, so I've just retrieved what I can.

#### 2016-01

In [50]:
with open('html files/2016-01 366 Daily Inspirational Quotes for 2016, One Month at a Time.html') as f:
    soup = BeautifulSoup(f)

    # delete strong/br elements, most of holiday names are in strong
    # for element in soup(['strong', 'br', 'a']):
    #     element.decompose()
    
    # delete empty tags
    for tag in soup.find_all():
        if not tag.text.strip() and not tag.contents:
            tag.decompose()

    dates = soup.find_all('h2')
    
    quotes = []
    
    for idx, date in enumerate(dates):
        date_text = date.text.split("(")[0].strip()
        # %A = day of week full name e.g. Sunday
        # #B = month full name e.g. January
        # %d = day of month with zero padding
        # %-d = day as decimal number i.e. without zero padding
        # %Y = full year e.g. 2025
        date_format = "%A, %B %d"
        try:
            parsed_date = datetime.strptime(date_text, date_format).replace(year=2016).strftime("%Y-%m-%d")
        except ValueError:
            try:
                parsed_date = datetime.strptime(date_text, "%A %B %d, %Y").strftime("%Y-%m-%d")
            except ValueError:
                print(f"error: date format doesn't match:\n{date_text}")
                continue

        quote = date.find_next('p').next
        quote_text, author_text = quote.split("”")  # confirmed that all quotes have this char
        quote_text = quote_text.strip().lstrip("“")

        author_text = author_text.strip()[1:]

        quotes.append({"date": parsed_date, 
                            "quote": quote_text,
                            "author": author_text})

with open('quotes_2016-01.json', 'w') as f:
    json.dump(quotes, f, indent=4)

#### 2016-02

In [122]:
skip_list = [9, 21, 22]

with open('html files/2016-02 29 Smart, Funny, and Moving Quotes to Inspire You In February.html') as f:
    soup = BeautifulSoup(f)


    # delete strong/br elements, most of holiday names are in strong
    # for element in soup(['strong', 'br', 'a']):
    #     element.decompose()
    
    # delete empty tags
    for tag in soup.find_all():
        if not tag.text.strip() and not tag.contents:
            tag.decompose()

    dates = soup.find_all('h2')

    
    quotes = []
    
    for idx, date in enumerate(dates):
        if idx in skip_list:
            continue
        # split date string based on : and then ( 
        date_text = re.split(r"[:\(]", date.text, maxsplit=1)[0].strip()
        date_text += " 2016"
        # %A = day of week full name e.g. Sunday
        # #B = month full name e.g. January
        # %d = day of month with zero padding
        # %-d = day as decimal number i.e. without zero padding
        # %Y = full year e.g. 2025
        date_format = "%A, %B %d %Y"
        try:
            parsed_date = datetime.strptime(date_text, date_format).replace(year=2016).strftime("%Y-%m-%d")
        except ValueError as e:
            print(e)
            try:
                parsed_date = datetime.strptime(date_text, "%A %B %d, %Y").strftime("%Y-%m-%d")
            except ValueError:
                print(f"error: date format doesn't match:\n{date_text}")
                continue
            
        quote = date.find_next('p').next
        quote_text = quote.text.strip().lstrip("“").rstrip("”")
        author_text = quote.find_next('p').next.text.strip().lstrip('—').strip()
    
        quotes.append({"date": parsed_date, 
                        "quote": quote_text,
                        "author": author_text})

# manually add quotes in skip_list
quotes.append({'date': '2016-02-10',
 'quote': 'My perfect date night: I pick you up. In my Kia Sorrento. You get in. There’s candles in the car. You go ‘Is that dangerous?” and I go, ‘Yes… but I like danger.’\nWe go to your favorite restaurant, and we have a fantastic meal. We come outside and we see my car’s on fire. You go, ‘Aziz, your car’s on fire. Aren’t you upset?’\nI pull out a bag of marshmallows and I go, ‘No. I knew this was gonna happen.’ And then I kiss you. In front of my burning car.',
 'author': 'Aziz Ansari'})

quotes.append({'date': '2016-02-22',
 'quote': 'If you text ‘I love you’ and the person writes back an emoji – no matter what that emoji is – they don’t love you back.',
 'author': 'Chelsea Peretti'})

quotes.append({'date': '2016-02-23',
 'quote': 'Love thy neighbor — and if he happens to be tall, debonair and devastating, it will be that much easier.',
 'author': 'Mae West'})


with open('quotes_2016-02.json', 'w') as f:
    json.dump(quotes, f, indent=4)

In [124]:
# skip_list = [9, 21, 22]

with open('html files/2016-03 31 Quotes About Making Your Own Luck for March 2016.html') as f:
    soup = BeautifulSoup(f)


    # delete strong/br elements, most of holiday names are in strong
    # for element in soup(['strong', 'br', 'a']):
    #     element.decompose()
    
    # delete empty tags
    for tag in soup.find_all():
        if not tag.text.strip() and not tag.contents:
            tag.decompose()

    dates = soup.find_all('h2')

    
    quotes = []


    


In [ ]:
if True:
    for idx, date in enumerate(dates):
        if idx in skip_list:
            continue
        # split date string based on : and then ( 
        date_text = re.split(r"[:\(]", date.text, maxsplit=1)[0].strip()
        date_text += " 2016"
        # %A = day of week full name e.g. Sunday
        # #B = month full name e.g. January
        # %d = day of month with zero padding
        # %-d = day as decimal number i.e. without zero padding
        # %Y = full year e.g. 2025
        date_format = "%A, %B %d %Y"
        try:
            parsed_date = datetime.strptime(date_text, date_format).replace(year=2016).strftime("%Y-%m-%d")
        except ValueError as e:
            print(e)
            try:
                parsed_date = datetime.strptime(date_text, "%A %B %d, %Y").strftime("%Y-%m-%d")
            except ValueError:
                print(f"error: date format doesn't match:\n{date_text}")
                continue
            
        quote = date.find_next('p').next
        quote_text = quote.text.strip().lstrip("“").rstrip("”")
        author_text = quote.find_next('p').next.text.strip().lstrip('—').strip()
    



In [126]:
quotes

[{'date': '2016-03-01',
  'quote': 'No one is going to know or care about your failures, and neither should you. All you have to do is learn from them and those around you. [A]ll that matters in business is that you get it right once. Then everyone can tell you how lucky you are.',
  'author': '–Mark Cuban'},
 {'date': '2016-03-02',
  'quote': 'Shallow men believe in luck or in circumstance. Strong men believe in cause and effect.',
  'author': '?Ralph Waldo Emerson'},
 {'date': '2016-03-03',
  'quote': 'You never know what worse luck your bad luck has saved you from.',
  'author': '?Cormac McCarthy,'},
 {'date': '2016-03-04',
  'quote': 'Luck is a dividend of sweat. The more you sweat, the luckier you get.',
  'author': '–Ray Kroc'},
 {'date': '2016-03-05',
  'quote': 'Luck can only get you so far.',
  'author': '–J. K. Rowling'},
 {'date': '2016-03-06',
  'quote': 'Hard work and a proper frame of mind prepare you for the lucky breaks that come along — or don’t.',
  'author': '-Harris

In [ ]:
        quotes.append({"date": parsed_date, 
                        "quote": quote_text,
                        "author": author_text})

with open('quotes_2016-03.json', 'w') as f:
    json.dump(quotes, f, indent=4)

### Combining all the quotes together

In [9]:
list("quotes_2016-01.json quotes_2016-02.json quotes_2017.json quotes_2018.json quotes_2019.json quotes_2020.json quotes_2022.json quotes_2023.json quotes_2024.json quotes_2025.json".split(" "))

['quotes_2016-01.json',
 'quotes_2016-02.json',
 'quotes_2017.json',
 'quotes_2018.json',
 'quotes_2019.json',
 'quotes_2020.json',
 'quotes_2022.json',
 'quotes_2023.json',
 'quotes_2024.json',
 'quotes_2025.json']

In [11]:
json_files = [
    'quotes_2016-01.json',
    'quotes_2016-02.json',
    'quotes_2017.json',
    'quotes_2018.json',
    'quotes_2019.json',
    'quotes_2020.json',
    'quotes_2022.json',
    'quotes_2023.json',
    'quotes_2024.json',
    'quotes_2025.json'
]

merged_quotes = []

for file in json_files:
    try:
        with open(file, "r") as f:
            quotes = json.load(f)
            merged_quotes.extend(quotes)
    except (FileNotFoundError) as e:
        print(f"Skipping {file}: {e}")
        continue

with open("quotes.json", "w") as f:
    json.dump(merged_quotes, f, indent=2)

### Split author from quote text

> **NOTE:** not used anymore because it's easier to use `bs4` to get the author directly if it's in a tag like `<br>` or `<em>` and in these strings the separator is not always the same, different types of dashes 

In [3]:
df = pd.read_json('quotes.json')
df.head()

Date                                              Quote
0 2024-01-01  “Life becomes easier when you learn to accept ...
1 2024-01-02  “If you don’t fall, how are you going to know ...
2 2024-01-03  “You are in danger of living a life so comfort...
3 2024-01-04  “This life is yours. Take the power to choose ...
4 2024-01-05  “Take the opportunity to learn from your mista...

In [4]:
print(ord('—'))

8212


In [5]:
df['count 8212'] = df['Quote'].str.count(chr(8212))

In [6]:
# some have three...
df['count 8212'].describe()

count    1455.000000
mean        0.852921
std         0.389446
min         0.000000
25%         1.000000
50%         1.000000
75%         1.000000
max         3.000000
Name: count 8212, dtype: float64

In [28]:
df[df['count 8212'] == 1]['Quote'].str.split(chr(8212))[0]

['“Life becomes easier when you learn to accept the apology you never got.”',
 'R. Brault']

In [52]:
# get author from quote
df[['quote8212', 'author8212']] = df['Quote'].str.rsplit(chr(8212), n=1, expand=True).iloc[:, :2]

In [59]:
# These correctly output because of str.rsplit with n=1
df[df['count 8212'] > 1]

Date                                              Quote  \
771  2022-02-17  “You do have a story inside you; it lies artic...   
833  2022-04-20  “Far better is it to dare mighty things, to wi...   
836  2022-04-23  “Luck? I don’t know anything about luck. I’ve ...   
863  2022-05-20  “Love is our true destiny. We do not find the ...   
953  2022-08-18  “Even more, I had never meant to love him. One...   
975  2022-09-09  “Don’t take too much advice. Most people who h...   
1003 2022-10-07  “It’s important to be precise about words, bec...   
1009 2022-10-13  “I don’t look to jump over seven-foot bars — I...   
1032 2022-11-05  “Before you speak, run this through your head:...   
1105 2020-01-17  “Rarely are opportunities presented to you in ...   
1448 2020-12-25  “I hate that word — ‘lucky.’ It cheapens a lot...   

      count 8212                                          quote8212  \
771            2  “You do have a story inside you; it lies artic...   
833            3  “Far better is it to dare mighty things, to wi...   
836            2  “Luck? I don’t know anything about luck. I’ve ...   
863            2  “Love is our true destiny. We do not find the ...   
953            3  “Even more, I had never meant to love him. One...   
975            3  “Don’t take too much advice. Most people who h...   
1003           2  “It’s important to be precise about words, bec...   
1009           2  “I don’t look to jump over seven-foot bars — I...   
1032           2  “Before you speak, run this through your head:...   
1105           3  “Rarely are opportunities presented to you in ...   
1448           2  “I hate that word — ‘lucky.’ It cheapens a lot...   

                                author8212  
771                              Anne Rice  
833                     Theodore Roosevelt  
836                           Lucille Ball  
863                          Thomas Merton  
953                        Stephenie Meyer  
975   Ben Silbermann, founder of Pinterest  
1003                       Michael Nesmith  
1009                        Warren Buffett  
1032                         Beret Guidera  
1105                        Susan Wojcicki  
1448                        Peter Dinklage

In [60]:
# these have a different hyphen...
df[df['count 8212'] == 0]

Date                                              Quote  \
32   2024-02-02  “Man is free at the instant he wants to be.”–V...   
366  2023-01-01  “Life is a daring adventure or it is nothing a...   
369  2023-01-04  “Intuitively, you would think the more fairnes...   
370  2023-01-05  “Opportunities don’t happen. You create them.”...   
371  2023-01-06    “Nothing will work unless you do.”–Maya Angelou   
...         ...                                                ...   
722  2023-12-30  “It is amazing what you can accomplish if you ...   
723  2023-12-31  “The bad news is time flies. The good news is ...   
794  2022-03-12  “Before long, many people will telecommute fro...   
899  2022-06-25  “If you want one thing too much it’s likely to...   
1082 2022-12-25  “God never gives someone a gift they are not c...   

      count 8212                                          quote8212 author8212  
32             0  “Man is free at the instant he wants to be.”–V...       None  
366            0  “Life is a daring adventure or it is nothing a...       None  
369            0  “Intuitively, you would think the more fairnes...       None  
370            0  “Opportunities don’t happen. You create them.”...       None  
371            0    “Nothing will work unless you do.”–Maya Angelou       None  
...          ...                                                ...        ...  
722            0  “It is amazing what you can accomplish if you ...       None  
723            0  “The bad news is time flies. The good news is ...       None  
794            0  “Before long, many people will telecommute fro...       None  
899            0  “If you want one thing too much it’s likely to...       None  
1082           0  “God never gives someone a gift they are not c...       None  

[229 rows x 5 columns]

In [64]:
# ord("–") # 8211
df['count 8211'] = df['Quote'].str.count(chr(8211))

In [65]:
# get author from quote with different hyphen
df[['quote8211', 'author8211']] = df['Quote'].str.rsplit(chr(8211), n=1, expand=True).iloc[:, :2]
df

In [68]:
ord("—")

8212

In [76]:
# group them all together
df['quote by itself'] = df['quote8212']
df['author by itself'] = df['author8212'].fillna(df['author8211'])

In [77]:
df[df['quote by itself'].isna()]

Empty DataFrame
Columns: [Date, Quote, count 8212, quote8212, author8212, count 8211, quote8211, author8211, quote by itself, author by itself]
Index: []

In [78]:
df[df['author by itself'].isna()]

Empty DataFrame
Columns: [Date, Quote, count 8212, quote8212, author8212, count 8211, quote8211, author8211, quote by itself, author by itself]
Index: []

In [79]:
# doing visual inspection still comes up with quotes and author together?
df[['Date', 'quote by itself', 'author by itself']].to_csv('quotes.csv')

In [81]:
df.to_csv('quotes2.csv')

easier to do in excel/sheets by just copy and paste the 8211 quote in place of quote by itself for those 229 rows where author8212 is empty

In [62]:
# not an easy way to do it with the quotes
# df['count open quote'] = df['Quote'].str.count(chr(8220))
# df['count close quote'] = df['Quote'].str.count(chr(8221))
# df['count open quote'].sum(), df['count close quote'].sum(), df.shape[0]
# df = df.drop(columns=['count open quote', 'count close quote'])

In [ ]:
# fix what I can with code
# fix the rest by hand...

In [82]:
df = pd.read_csv('quotes3.csv')
df.head()

Unnamed: 0        Date                                              Quote  \
0           0  2024-01-01  “Life becomes easier when you learn to accept ...   
1           1  2024-01-02  “If you don’t fall, how are you going to know ...   
2           2  2024-01-03  “You are in danger of living a life so comfort...   
3           3  2024-01-04  “This life is yours. Take the power to choose ...   
4           4  2024-01-05  “Take the opportunity to learn from your mista...   

                                           quote8212          author8212  \
0  “Life becomes easier when you learn to accept ...           R. Brault   
1  “If you don’t fall, how are you going to know ...       Stephen Curry   
2  “You are in danger of living a life so comfort...       David Goggins   
3  “This life is yours. Take the power to choose ...  Susan Polis Schutz   
4  “Take the opportunity to learn from your mista...        Tony Robbins   

   count 8211                                          quote8211 author8211  \
0           0  “Life becomes easier when you learn to accept ...        NaN   
1           0  “If you don’t fall, how are you going to know ...        NaN   
2           0  “You are in danger of living a life so comfort...        NaN   
3           0  “This life is yours. Take the power to choose ...        NaN   
4           0  “Take the opportunity to learn from your mista...        NaN   

                                     quote by itself    author by itself  
0  “Life becomes easier when you learn to accept ...           R. Brault  
1  “If you don’t fall, how are you going to know ...       Stephen Curry  
2  “You are in danger of living a life so comfort...       David Goggins  
3  “This life is yours. Take the power to choose ...  Susan Polis Schutz  
4  “Take the opportunity to learn from your mista...        Tony Robbins

In [86]:
df['date'] = df['Date']
df['quote'] = df['quote by itself']
df['author'] = df['author by itself']

In [89]:
df[['date', 'quote', 'author']].to_json('quotes2.json', orient='records')